### PR TITLE
feat(ui): step 3 — sidebar sessions list + browser-style session tabs

### PR DESCRIPTION
**Pull request by winoooops**
## Summary

Step 3 of the UI handoff migration (roadmap §9). Restyles the sidebar sessions list per handoff §4.2 and replaces the cycle-2 placeholder strip with a real browser-style SessionTabs component per §4.3.

### Sidebar (§4.2)
- **Active vs Recent groupings**: running/paused → Active (drag-orderable), completed/errored → Recent (read-only list).
- **Row layout**: StatusDot + bold title + relative time on row 1; subtitle (currentAction or cwd basename) on row 2; state pill + +added/-removed deltas on row 3.
- **Active row**: lavender-tinted `bg-primary/10` with a 2px lavender accent bar on the left edge.
- Active and Recent groups share a single `min-h-0 flex-1 overflow-y-auto` scroll region (so a long Recent list can't push FileExplorer below the fixed sidebar).

### SessionTabs (§4.3)
- 38px strip with 30px tabs that lift via `-mb-px` so the active tab visually merges into the canvas below.
- Per-tab agent identity chip drives accent color from the `AGENTS` registry.
- Roving keyboard focus (ArrowLeft/ArrowRight cycle, Enter/Space activate) per the WAI-ARIA tabs pattern. `tabIndex=0` only on the active tab.
- Status pip + close X + ghost `+` button. Filter keeps `running`/`paused` sessions plus the active session even after its PTY exits (so a Restart pane keeps its tab).
- Closing the active tab pre-selects the next visible tab using the `open` order (not the full `sessions` index), so the strip's selection doesn't jump to a hidden Recent session.
- Replaces TerminalZone's legacy `tab-bar` (removed; obsolete now that SessionTabs drives the same state).

### New shared helpers
- `src/features/workspace/utils/agentForSession.ts` — maps `Session.agentType` → `AGENTS` registry entry (claude-code → claude, aider/generic → shell fallback).
- `src/features/workspace/components/StatusDot.tsx` — 7px rounded span with running/paused/completed/errored tone classes; reused by Sidebar rows and SessionTabs pips.

## Test plan
- [x] `npm run lint` — clean
- [x] `npm run type-check` — clean
- [x] `npm run test` — 1674 passed, 3 skipped
- [x] `npm run format:check` — clean
- [x] Codex review (6 cycles): every behavioral concern addressed (duplicate UI, exited active tab, scroll constraint, roving focus, child key bubble, layoutScroll node, visible-order close fallback, close-then-select ordering)
- [ ] Manual smoke: drag-reorder Active sessions; close active tab and verify visible next selected; ArrowLeft/Right cycles tabs; close X is reachable by keyboard; long Recent group scrolls within the shared region.

### CODE REVIEW

**Comment by winoooops** [comment]
**Addressed in 3b080f6** (codex-verify cycle 1: pass — _patch is correct, confidence 91%_).

- F1 (HIGH, errored mode): TerminalZone now routes `completed || errored` → `awaiting-restart`. Restart UX is reachable for both exited statuses.
- F2 (MEDIUM, ARIA tablist): split the boundary so `role="tablist"` wraps only the tabs; `+` button + spacer live outside.
- F3 (LOW, Active double-padding): dropped `px-3` from the Active group's wrapper flex; GroupHeader's own padding now matches the Recent header.

Also addressed two adjacent items in the same commit: codex-connector P2 (sidebar close fallback now mirrors SessionTabs.handleClose), and user feedback on the row `rounded-xl` → `rounded-[8px]` so the lavender accent reads as a clean vertical line.

Two follow-ups filed as out-of-scope: #175 (SESSIONS/FILES/CONTEXT top-tab switcher per §4.2) and #176 (double-scrollbar in FileExplorer area, pre-existing).

File: src/features/workspace/components/Sidebar.tsx [hunk 1]
```diff
@@ -149,7 +223,7 @@ const SessionItem = ({
             e.stopPropagation()
             onRemove?.(session.id)
           }}
```
[the remark this answers is not in the document]
Comment (winoooops) [reply src/features/workspace/components/Sidebar.tsx]: **Addressed in 3b080f6** (codex-verify cycle 1: pass).

Sidebar's `handleRemoveSession` now mirrors `SessionTabs.handleClose`: when the active session is removed and there are other Active rows, pre-select the next visible Active row before delegating to `onRemoveSession`. Active is the running+paused subset, so the manager's full-sessions fallback can no longer land on a sandwiched Recent (completed/errored) row.

**Comment by winoooops** [comment]
**Addressed in cd476a7** (codex-verify cycle 2: pass — _patch is correct, confidence 91%_).

- F1 (HIGH, call-order regression): inverted to remove-then-override, matching the established `SessionTabs.handleClose` pattern. Cycle-1's fix is now actually doing what its comment claimed.
- F2 (MEDIUM, auto-activation): `focusTabAtOffset` no longer calls `onSelect` — Arrow keys move DOM focus only, Enter/Space activates (manual-activation per WAI-ARIA tabs).
- F3 (LOW, tabpanel linkage): added matching `id` + `aria-controls` on tabs and `id` + `role="tabpanel"` + `aria-labelledby` on the TerminalZone pane wrapper, both keyed by `session.id`.

Tests rewritten: arrow-key assertions now check focus movement (not onSelect calls); new test covers Enter activating a focused inactive tab.

**Comment by winoooops** [comment]
**Addressed in 35f04ec** (codex-verify cycle 3: pass — _patch is correct, confidence 90%_).

- F1 (MEDIUM, sidebar/strip divergence): extracted shared `pickNextVisibleSessionId(sessions, removedId, activeId)` helper into `utils/`. Builds the candidate set as `running/paused ∪ {activeId}` (matching SessionTabs.open semantics) so an exited-active removal picks the same neighbor whether the user clicks remove from the sidebar or the close X in the tab strip. Both sites now route through the helper.
- F2 (LOW, Recent props): narrowed RecentSessionRow to `Omit<SessionRowProps, 'onRename'>` so the read-only contract is enforced by TypeScript.

7 new helper tests cover the edge cases (right-neighbor, left-wrap, hidden Recents skipped, exited-active included, single-visible returns undefined).

**Comment by winoooops** [comment]
**Addressed in 3753ea59e17eed6b70ad9af1d29b4170bfda705e** (codex-verify cycle 4: pass — _patch is correct, confidence 93%_).

- F1 (MEDIUM, close button roving): added `tabIndex={isActive ? 0 : -1}` to the close <button>. Roving now extends to interactive descendants per WAI-ARIA tabs — Tab cleanly skips over inactive tabs' close buttons.
- F2 (LOW, optional handler): `handleRemoveSession` is now `undefined` when `onRemoveSession` is omitted. RecentSessionRow's `{onRemove && <button>}` guard hides the button correctly; no more silent select-without-remove.

New regression tests cover both: inactive close button has `tabindex="-1"`; sidebar without onRemoveSession does not render the Recent remove button.

**Comment by winoooops** [comment]
**Addressed in 1b1d2ca80f81876de3ee35ca2e814841b431af3d** (codex-verify cycle 5: pass — _patch is correct, confidence 90%_).

- F1 (MEDIUM, focus loss): `queueMicrotask` after `onSelect(nextId)` calls `document.getElementById('session-tab-${nextId}')?.focus()`. Keyboard users no longer get dropped to <body> after closing the active tab.
- F2 (MEDIUM, name pollution): explicit `aria-label={session.name}` on the role=\"tab\" div. Descendants (close button, StatusDot) keep their own labels for individual focus stops; the tab's announced name stays clean.
- F3 (LOW, stale closure): `onReorder` recomputes the Recent partition from the latest `sessions` prop inside the callback — no more drop / mis-placement when an agent completes a task mid-drag.

Plus three user-feedback items addressed in the same commit: Recent row dim treatment (StatusDot hollow border + dim state pills + dimmer text), hover-time hiding so action buttons don't overlap, and rename support added to RecentSessionRow (with edit/rename input + hover edit button).

**Comment by winoooops** [comment]
**Addressed in bc51bd71d5a5110ecdd12b35278d2687d78757b2** (codex-verify cycle 6: pass — _patch is correct, confidence 91%_).

- F1 (HIGH, <input> in <button>): restructured both SessionRow and RecentSessionRow into an absolute click-button background + a pointer-events-none foreground overlay. The rename input is now a sibling of the activation button, not a descendant — no more HTML spec violation, no more Firefox/Safari focus quirk. Interactive bits (input, hover buttons, double-clickable title) opt back into pointer events with `pointer-events-auto`.
- F2 (MEDIUM, <div> in <ul>): `<div data-testid="active-empty">` → `<li data-testid="active-empty">` so the empty state is a valid `Reorder.Group` child.
- Codex re-verify caught a regression introduced by the F1 restructure: title clicks no longer activated the row because the title span is now a sibling of the activation button instead of a descendant. Added explicit `onClick={() => onSessionClick(session.id)}` on the title spans in both rows; `cursor-pointer` makes the affordance visible. Double-click rename still works.

**Comment by winoooops** [comment]
**Addressed in 2b001a521884e026dc7db2807ef250b1f46d0709** (codex-verify cycle 7: pass — _patch is correct, confidence 93%_).

- F1 (MEDIUM): SessionRow's edit + remove buttons now wrapped in `{onRename && (…)}` / `{onRemove && (…)}` to match the guards already present on RecentSessionRow. No more non-functional rename input when `onRenameSession` is omitted by the consumer.

Two LOW findings deferred:
- F2 (tablist Arrow keys): intentional gap — see the SessionTabs STUB comment + #177.
- F3 (SessionRow / RecentSessionRow duplication): exactly the dedup work scoped in #178 (filed earlier this round, lifts `SessionCard` + `SessionGroup` primitives out of Sidebar).

**Comment by winoooops** [comment]
**Addressed in 77db689962fba81c245d12a56408efa22e1db08b** (codex-verify cycle 8: pass — _patch is correct, confidence 91%_).

- F1 (LOW, SessionRow rename guard): backported `if (!onRename) return` from RecentSessionRow's onDoubleClick handler. Cycle-7 added the button guard but missed this twin path.
- F2 (LOW, dead aria-labelledby): TerminalZone now omits aria-labelledby on hidden panes for completed/errored non-active sessions (no corresponding tab exists in the strip). axe-core would flag those broken refs on CI.

Note on overall verdict: Claude marked this round as **✅ patch is correct**. Continuing the loop one more time to confirm the new commit comes back clean.

**Comment by winoooops** [comment]
**Addressed in 311aa0ab6175b468f85d46acb77c0cb05b95131d** (codex-verify cycle 9: pass — _patch is correct, confidence 95%_).

- F1 (MEDIUM, double announcement): `aria-hidden="true"` on the visible name span in both SessionRow and RecentSessionRow. Overlay button's `aria-label` is now the sole source of the row's announced name; subtitle, state pill, and line-delta stay traversable for context. Rename input keeps its own `aria-label="Rename session"`.

LOW findings deferred to follow-ups (already filed):
- Arrow-key cycling on tablist → #177 (global keybinding work).
- Editing-state duplication between SessionRow / RecentSessionRow → #178 (Sidebar refactor + extract `useSessionRowEditing` hook).

**Comment by winoooops** [comment]
**Addressed in db5778f08259aebe53e5fdc8fbb1a494de5ca64f** (codex-verify cycle 10: pass — _patch is correct, confidence 90%_).

- F1 (MEDIUM, null-active tabindex): new `isFocusEntryPoint` prop drives both the tab div and the close button. Equals `isActive` in steady state; falls back to the first visible tab when `activeSessionId` is null so keyboard users always have an entry point. New test covers the null case explicitly.
- F3 (LOW, ordering assertion): added `expect(onClose.mock.invocationCallOrder[0]).toBeLessThan(onSelect.mock.invocationCallOrder[0])` to the visible-order close test. The flushSync invariant is now enforced by tests, not just by the comment.

LOWs deferred:
- Editing-state duplication → #178.
- Tablist ArrowLeft/Right → #177.

Note on overall verdict: Claude marked this round as **✅ patch is correct**. Continuing one more cycle to confirm clean on the new commit.

**Comment by winoooops** [comment]
**Addressed in 6d059e5b7612a4ae98e2eaa150c76964a7306e05** (codex-verify cycle 11: pass — _patch is correct, confidence 89%_).

- F1 (MEDIUM, single Tab stop): close button now `tabIndex=-1` always; keyboard close via Delete/Backspace on the focused tab (browser-tab Cmd+W convention). Tab from the active tab now falls through to the tabpanel cleanly per WAI-ARIA §3.27.
- F2 (MEDIUM, rename duplication): extracted `useRenameState(session, onRename)` hook with `{isEditing, editValue, setEditValue, inputRef, beginEdit, commitRename, cancelRename}`. Both SessionRow and RecentSessionRow consume the hook — the rename contract is now in one place. 8 new hook tests cover begin/commit/cancel + edge cases.
- F3 (LOW, e2e helper name): `getSessionLabel` → `getLatestSessionTabId`. Function name matches what it returns.

This addresses the most pressing piece of the larger Sidebar refactor scoped in #178 — the rename contract. Full structural move (Sidebar to global components, SessionCard variant primitive) still belongs in #178.

**Comment by winoooops** [comment]
**Addressed in 28df6df0eed45dddd1f3b8b493c47df95a9b7f83** (codex-verify cycle 12: pass — _patch is correct, confidence 93%_).

- F1 (MEDIUM, keyboard a11y): added `group-focus-within:opacity-100` to both row action overlays. Tab-focusing the row overlay now reveals rename/remove — keyboard users finally have a path to rename.
- F2 (MEDIUM, dedup visible-filter): extracted `getVisibleSessions(sessions, activeSessionId)` from `pickNextVisibleSessionId.ts`. Both SessionTabs and the close-fallback util consume it; the visible-set definition can't silently drift between render and navigation.
- F3 (LOW, comment): queueMicrotask comment now correctly states focus restoration fires for both mouse and keyboard closes.

**Comment by winoooops** [comment]
**Addressed in 18825e25056c49e131dce08352a885e380b60b2e** (codex-verify cycle 13: pass — _patch is correct, confidence 88%_).

- F1 (MEDIUM, commit double-fire): `committedRef` guard inside `useRenameState`. `commitRename` early-returns once committed; `beginEdit` + `cancelRename` reset for the next rename session. The trailing blur after Enter is now a no-op. Two new tests lock the contract.
- F2 (LOW, sidebar focus): `handleRemoveSession` now `queueMicrotask`s a focus call to the new active row's overlay button after `onSessionClick(nextId)`. Mirrors SessionTabs.handleClose §4.4.3.

**Comment by winoooops** [comment]
**Addressed in 87b802fc9fdebc9e28b9558bc042c1143cd185dc** (codex-verify cycle 14: pass — _patch is correct, confidence 92%_).

All four LOW polish items addressed:
- F1: TerminalZone empty-state copy points at the SessionTabs strip above (the + lives there now).
- F2: removed misleading "always current" comment; `recentGroup` and `sessions` are equivalent render-time captures, so use the precomputed group directly.
- F3: `data-role="activate"` on the overlay button + selector update so the focus-restoration target survives DOM reordering.
- F4: imported `isOpenSessionStatus`; Active = the predicate, Recent = its complement — future non-open statuses auto-land in Recent.

Note: Claude marked this round as **✅ patch is correct**. One cycle remaining of the 15-cap.

**Comment by winoooops** [comment]
**Cycle 15 — addressed in a3c56d00cca2fcbeea243249c7d567607d35c7c9** (codex-verify pass — _patch is correct, confidence 96%, 0 new issues_).

Took your preferred Option 1: both `SessionRow` and `RecentSessionRow` overlay `<button>`s now render with `id={`sidebar-activate-${session.id}`}`, and `Sidebar.handleRemoveSession` switched to `document.getElementById(`sidebar-activate-${nextId}`)`. The two parallel focus-restoration paths (Sidebar + SessionTabs) now use the same lookup mechanism — fewer ways for future changes to drift. `data-role="activate"` is preserved as a stable hook (no callers depend on it today, but it's harmless).

Inline comment rewritten to call out the actual reason this matters: id-based lookup keeps session ids out of the CSS-attribute-selector parser, so a future schema change that admits `"` or `]` won't break focus restoration silently or throw `SyntaxError` from inside `queueMicrotask`.

Pattern KB: appended finding #14 to `docs/reviews/patterns/accessibility.md` capturing the "interpolation into selector parser" class so the same heuristic can be applied next time.

Tests: 1700 passed, 3 skipped. Lint + type-check clean. **This is the final cycle of the 15-cycle review cap** — PR is now both Claude- and Codex-clean (Codex P2 thread on arrow-key nav was wontfix-resolved earlier this cycle, tracked in #177).

**Comment by winoooops** [comment]
**Cycle 16 — addressed in 94f79f801ab9cd61191b1264358ee1816843f028** (codex-verify pass — _patch is correct, confidence 94%, 0 new issues_).

All three findings:

- **F1 (MEDIUM, _cancelRename resets committedRef_):** `cancelRename` now sets `committedRef.current = true` (single-character fix per your IDEA-A). The "next rename session starts clean" invariant stays — `beginEdit` already unconditionally resets the guard. **Added the missing regression test** that simulates the Escape→stale-onBlur path (calls `cancelRename()` then `commitRename()` and asserts `onRename` was NOT invoked) — this would have caught the bug at static analysis time but obviously didn't because jsdom doesn't fire native blur on DOM removal.
- **F2 (LOW, _e2e raw ${id} CSS selector_):** swapped `getLatestSessionTabId` to `document.getElementById(`session-panel-${id}`)` — `TerminalZone` already renders that id (added in earlier cycles for `aria-labelledby`). Same pattern as cycle 15's Sidebar fix; you correctly noticed the asymmetry.
- **F3 (LOW, _exited tab silent on status change_):** `SessionTab` `aria-label` now appends ` (ended)` for completed and errored statuses; running/paused unchanged. Test asserts both exited variants. WCAG 2.1 SC 4.1.3 satisfied without changing the visual heartbeat-only-for-live-sessions language.

Bonus: also addressed Codex's P2 inline finding (`sessionSubtitle` Windows separator normalization) in the same cycle, and applied a user-directed widening from "basename only" → "last 2 segments" so `/home/will` reads as `home/will` instead of just `will`. Plus a user-flagged Recent-row hover overlay fix (the right-aligned subtitle was visually overlapping the absolute hover buttons; added `group-hover:opacity-0` to match the existing timestamp pattern).

Pattern KB: appended findings #9 (`react-lifecycle.md` — guard-ref must be armed by both commit AND cancel paths), #15 (`accessibility.md` — production fix scope should propagate to test infra in the same cycle), and #16 (`accessibility.md` — silent retention is an a11y regression even when the element exists). Index counts bumped.

Tests: 1704 passed, 3 skipped (added 4 regression tests). Companion CI fix `ea95d94` (e2e pty-spawn poll) landed earlier in the session as a non-trailer commit per the upsource-review Step 1.5 protocol.

**Comment by winoooops** [comment]
**Cycle 17 — addressed in c5f80e4c40dd47323d65534b14e3a549eefcf5c3** (codex-verify pass — _correct, confidence 94%, 0 new issues, 0 deferred_).

All three findings:

- **F1 (MEDIUM, _hasVisibleTab duplicates getVisibleSessions predicate_):** TerminalZone now imports `isOpenSessionStatus` from `../utils/pickNextVisibleSessionId` and computes `hasVisibleTab` as `isActive || isOpenSessionStatus(session.status)`. Future status extensions auto-flow into both visibility surfaces.
- **F2 (MEDIUM, _isFocusEntryPoint stale-activeSessionId fallback_):** added `hasFocusMatch = open.some(s => s.id === activeSessionId)` once at the SessionTabs body. Condition is now `id === activeSessionId || (!hasFocusMatch && idx === 0)` — collapses three roving-focus cases (initial null, fresh-load no-active, stale flushSync) to one first-tab fallback. Regression test added that passes a non-null-but-stale `activeSessionId` and asserts the first visible tab carries `tabIndex=0`. Closes the WAI-ARIA invariant under the flushSync intermediate render.
- **F3 (LOW, _onReorder stale recentGroup closure_):** added `recentGroupRef` synced via render-body assignment (`recentGroupRef.current = recentGroup` runs every render — the standard "always-latest" ref pattern, cheaper than `useEffect` and synchronous). `onReorder` reads `.current` instead of the closure-captured slice; Framer Motion's retained callbacks now see fresh values across mid-drag re-renders.

Pattern KB: appended findings #3 to `module-boundaries.md` (predicate inline-duplication is a coupling smell with functional consequence — same finding-class as the export-shape mismatch in #2), #17 to `accessibility.md` (roving-focus fallbacks must key on visible-set property not raw-state property), and #10 to `react-lifecycle.md` (callbacks registered with third-party machinery must read mutable state through refs).

Tests: 1705 passed, 3 skipped (added 1 regression test for F2). Cycle 17 of cap 20.

**Comment by winoooops** [comment]
**Already addressed in c5f80e4c40dd47323d65534b14e3a549eefcf5c3** (cycle 17, codex-verify pass — _correct, confidence 94%_).

This review ran against `94f79f8` (cycle 16), which still had the inline predicate. The exact finding was independently raised by your previous review (id 4393441052) and fixed in cycle 17 — see [issue-comment 4393585218](https://github.com/winoooops/vimeflow/pull/174#issuecomment-4393585218) for the cycle-17 reply. `TerminalZone.tsx` now imports `isOpenSessionStatus` and computes `hasVisibleTab` as `isActive || isOpenSessionStatus(session.status)`. No further action needed for this cycle.

The "✅ patch is correct" verdict on the rest of the PR is appreciated — agree on the WAI-ARIA tabs implementation, `pickNextVisibleSessionId` consolidation, and `useRenameState` double-fire guard.

**Comment by winoooops** [comment]
**Cycle 20 (final) — partial address in aced283eb47e458a80b0ab2c1672a0e6c1f590ba** (codex-verify pass — _correct, confidence 93%, 0 new issues_).

- **F2 LOW (sessionSubtitle empty-string fallback) — FIXED**. Changed the zero-parts branch from `return session.workingDirectory` to `return session.workingDirectory || '~'`. POSIX root `/` is preserved (truthy → falls through `||`); only the actual race-window empty-string case now falls back to `~`. Comment expanded to spell out the race-window rationale. Added regression test asserting an empty-cwd session renders `~` in its subtitle (scoped to the row via `within(getByTestId('session-row'))` to avoid colliding with the SidebarStatusHeader's separate `~`).
- **F1 MEDIUM (arrow-key tab navigation, WAI-ARIA non-conformance) — SKIPPED, wontfix per [#177](https://github.com/winoooops/vimeflow/issues/177)**. This is a re-raise of cycle 15's Codex P2, which was wontfix'd at user direction in [comment 4391061890](https://github.com/winoooops/vimeflow/pull/174#discussion_r3196846306). The user explicitly removed the in-component handler in cycle 5 (rationale: "the keydown events seem to not take effect when I press them") and tracks the global keybinding (`Cmd+Shift+]` / `Cmd+Shift+[`) under #177.

  **Surfacing your stronger argument to the user:** unlike the cycle-15 Codex finding, you correctly note that focus can reach the tablist via paths *other* than xterm — clicking a tab releases focus to the tab div, and Tabbing through the New Session button can land on the strip. So a tablist-level handler at `SessionTabs` (with `open` already in scope and `getElementById('session-tab-' + targetId)` for focus dispatch) WOULD be reachable for non-mouse users in those scenarios. This is additive — it does not conflict with #177's global keybinding for the in-terminal case. Not landing it unilaterally because the user's prior decision is recent and explicit; flagging here so they can decide.

This concludes the user-extended 5-cycle batch (cycles 16-20). Both reviewers green on the actively fixed items; #177 is the open follow-up for the optional in-strip handler.
